### ISSUE TITLE
Standardize usage of machine-wide caches

### ISSUE THREAD

**Issue by lihaoyi** [com-lihaoyi/mill#3930]
In the past we only had Coursier managing its own machine-global cache for maven central, but this is starting to grow:

- Custom Java version support needs somewhere to put the JVMs
- Android support needs somewhere to put the Android SDK and other heavyweight tools
- The example PythonModule needs somewhere to cache its PIP downloads
- The example TypescriptModule could use somewhere to cache its NPM downloads
- Once the filesystem-independent out/ folder layout lands, we could begin sharing out/ folder results between projects, and that would also need some standard place to put the artifacts

The basic idea is that when someone calls `clean`, they usually dont want to start all the way from scratch re-downloading every jar from Maven Central and re-downloading their JVM. These external dowloads typically are downloaded once and cached forever, and only very rarely do people want to clean them, in comparison to needing to clean local build outputs.

We should try and standardize how these "global" cached downloads are handled so people can just plug into the standard, rather than creatively coming up with their own solutions that end up being half-baked or inconsistent

**Comment by 0xnm** [comment]
My 2c on this: such cache structure vary a lot between the different ecosystems, because the particular ecosystem dictates the structure.

For example, in case of Maven dependencies the standard file cache layout supports artifact versioning as a part of the artifact's full path and it is easy to have different artifact versions out-of-project, somewhere on the machine.

If we take a look on the NPM dependencies, for example, `node_modules` doesn't support hosting different versions of the same artifact by default (unless package-aliasing is used? or maybe workspaces can help?). In this case such cache unlikely can be machine-wide and then becomes more project-wide (especially if we think that some dependencies are built-in, e.g. test reporting in Kotlin/JS, and versions can be different between different Mill installations). Now, let's say we have a project with several JS modules, and it won't be very wise to download the same NPM dependency several times (for each module), so it probably makes sense to have some folder at the root of the project, but there is no:

* built-in API to access root module (or root `out` folder). The closest way to access it (instead of using `T.dest` which is bound to the task) with the current API will be, probably, something like:

```
var m: mill.define.Module = this
while (m.millOuterCtx.enclosingModule != null) {
  m = m.millOuterCtx.enclosingModule.asInstanceOf[mill.define.Module]
}
val dir = m.millModuleBasePath.value / "out" / "js"
```

* way to control that there is only one job to download the particular dependency instead of the parallel ones. For example, if `Module A` declares a dependency `foo` and `Module B` declares the same dependency `foo`, it shouldn't be parallel downloads: if `Module A` is already downloading `foo`, `Module B` should wait and consume the result instead of shooting its own download job (is it the case currently with Maven dependencies download?).

**Comment by alexarchambault** [comment]
> * if `Module A` is already downloading `foo`, `Module B` should wait and consume the result instead of shooting its own download job (is it the case currently with Maven dependencies download?)

For coursier (that handles Maven dependencies), it's fine / safe to have several tasks or threads attempt to download the same file. Only one of them will effectively download the file, while the others wait for it (and may also download other things on their own in parallel), and another might take over if the first doing the download is interrupted (resuming the download if the repository supports it, not re-doing it from zero).

**Comment by alexarchambault** [comment]
AFAIK, Nix has a quite generic cache mechanism too, except it uses a checksum as key rather than a URL

**Comment by alexarchambault** [comment]
And Bazel must offer similar features too

**Comment by lihaoyi** [comment]
I think it would make sense to offer this as a built in feature using Coursier. 

* Keying on URLs is generally fine, since [https://www.w3.org/Provider/Style/URI](Cool URLs Don't Change)
* Coursier naturally has to provide all the bells and whistles around concurrency, interruption, etc. that would be a lot of work to re-implement in Mill.
* We can generally be confident in the robustness of the coursier implementation, since it is heavily exercised via dependency resolution in Mill's build and test suite, more so than any ad-hoc implementation we come up with

I think to make this a reality, a few steps are necessary:

1. Write reference documentation for programmatic use of the coursier cache API somewhere. https://get-coursier.io/docs/cache only covers the behavior of the cache but not how to actually use it from code. We could add the API usage docs there, or on mill-build.org, but it needs to live somewhere

2. Decide on whether to expose Coursier's API as-is or wrap it in a Mill wrapper helper. This depends largely on how well Coursier's Cache API fits the com-lihaoyi style that Mill and it's libraries are written in (simple function calls and objects, lots of default parameters, minimal builder or factory or config objects)

3. Review and discuss the exposed API and semantics to make sure it is what we want it to look like. Stable APIs in Mill are restricted from changing, so we should get it right the first time rather than needing to wait a year for the next breaking release to fix issues that could have been prevented up front

4. Write an example test in `example/fundamentals/libraries/` with documentation and include it on `fundamentals/bundled-libraries.adoc` together with usage docs for other bundled libraries

5. Update the `example/javalib/dependencies/4-downloading-unmanaged-jars` examples to use this new API rather than `requests.get` directly

It's a bunch of work but seems straightforward. Let's hold off on https://github.com/com-lihaoyi/mill/pull/4066 until we've done this, and then we can dogfood our new APIs to download the test repos rather than doing it ad-hoc.

@alexarchambault probably makes sense for you to take the lead on this whenever you have some spare cycles

**Comment by jodersky** [comment]
I'd like to add a feature request to this: allow arbitrary postprocessing of a file before caching it. For example, if I download a zip archive, I'd like to unzip it before caching the result.

**Comment by jodersky** [comment]
I'm just thinking about this in general, and this came to mind.

If we had reproducible `out/` folders (#4065), would it make sense to allow the results to be cached in a global directory? We could do this optionally by introducing a `Task.Global{}` (or something like that), where the result would be cached in a well-known location, under the result's checksum, and the `dest` folder would also be preserved there. We'd need some locking mechanism as well, but I think that this way we could share results "purely" with mill. It would also be a step in the direction of supporting remote caches.

**Comment by sequencer** [comment]
fwiw, we are using mill+nix in our project, see a minimal example here: https://github.com/chipsalliance/chisel-nix/blob/6a8222ad27b3fa4f2cd871152696afdc636c5b44/templates/chisel/nix/pkgs/mill-builder.nix#L6

**Comment by jodersky** [comment]
I agree, that's why I was thinking of specifically marking tasks somehow that would be suited for global caching. Maybe this mechanism could then also be used to mark which tasks should be cached remotely, when we add that feature.

**Comment by lihaoyi** [comment]
yeah it's possible. The remote caching POC (https://github.com/com-lihaoyi/mill/pull/2777) used a Mill query to select which tasks you wanted to remote cache, but there are other ways we could annotate it as well. Now we've started using `Task(foo = true){ ... }` to pass flags to tasks, we could use that approach to mark remote cacheable tasks in code rather than separately from the CLI

**Comment by sequencer** [comment]
I personally would like to seek for an option in mill that not downloading anything(sandboxing) but exposing the dependency to a machine-readable way that ask external tool to download and put them in a correct folder. So that we can safely cache it, and reuse in different clean machines.

The current state is that mill always assumes it has the network connection, which is far way from sandboxing. Sometimes ivy downloading even are not reproducible at all, and we need to cache that with an input hash see [here](https://github.com/chipsalliance/chisel-nix/blob/6a8222ad27b3fa4f2cd871152696afdc636c5b44/templates/chisel/nix/gcd/gcd.nix#L47)

Eventually what we expect is an `out` directory that only depends on the `build.sc`, and we use nix to take `build.sc` as input. For the dependencies in maven, we'd like to ask mill to dump the entire dependencies and we use coursier to download and cache each piece of them, then exposing the nix cache directory to mill.

**Comment by sideeffffect** [comment]
@jodersky 
> allow arbitrary postprocessing of a file before caching it. For example, if I download a zip archive, I'd like to unzip it before caching the result

I think I have good news regarding this. Among the many cool things Coursier can do, it downloads/"installs" various JVM distributions. And those are certainly distributed as zip or other archives. So it must be possible already today.

https://get-coursier.io/docs/cli-java

**Comment by lihaoyi** [comment]
One thing we may need to be concerned about for post-processing is where this post-processing is defined, how it is keyed for uniqueness, and how error-prone it is. The nice things about the dumb "download URL" workflow is that there is not much config to think about, it gives a great cache key to start with (the URL itself), it rarely fails (and can always be retried). 

Technically the downloaded contents can vary based on HTTP headers and all that (e.g. around authentication), but generally downloading stuff from public URLs is pretty straightforward

If we start allowing arbitrary post-processing, we need some way to hash the post-processing logic to add it to the cache key (maybe using `methodCodeHashSignatures`?), and we need to ensure the post-processing is appropriately configurable (I guess if it's arbitrary code that's already the case?) and that it doesn't fail or cause the caches to get into a bad state. Probably solvable problems, but definitely things we need to be careful about if we choose to go beyond just "dumb download URL" approach to the global caches

**Comment by jodersky** [comment]
I think we can indeed defer *arbitrary* postprocessing to a later point
(maybe once we have a robust solution to global/remote caching of tasks).
However, I would argue that archive extraction is a common enough scenario
to support.

**Comment by lihaoyi** [comment]
Apart from the download (and maybe unpack) kinds of caching, we also need some story for what to do with external tools outside of our control which require cache locations. PIP and NPM are two examples we're facing right now, but I'm sure there'll be others

**Comment by alexarchambault** [comment]
> I'd like to add a feature request to this: allow arbitrary postprocessing of a file before caching it. For example, if I download a zip archive, I'd like to unzip it before caching the result.

As @sideeffffect mentioned, this is already supported by coursier. Support for that has been added some time ago, for JVM handling support.

From the command-line, one can pass `--archive` to `cs get` to download and extract an archive:
```text
$ cs get https://cosmo.zip/pub/cosmocc/cosmocc-3.9.7.zip
~/Library/Caches/Coursier/v1/https/cosmo.zip/pub/cosmocc/cosmocc-3.9.7.zip

$ cs get https://cosmo.zip/pub/cosmocc/cosmocc-3.9.7.zip --archive
~/Library/Caches/Coursier/arc/https/cosmo.zip/pub/cosmocc/cosmocc-3.9.7.zip

$ ls "$(cs get https://cosmo.zip/pub/cosmocc/cosmocc-3.9.7.zip --archive)"
LICENSE.gpl2        LICENSE.lgpl3       bin                 libexec
LICENSE.gpl3        README.md           include             x86_64-linux-cosmo
LICENSE.lgpl2       aarch64-linux-cosmo lib
```
(note the `arc` instead of `v1` in the path)

From the API, one needs to use `coursier.cache.ArchiveCache`:
```scala
import coursier.cache.ArchiveCache
import coursier.util.Artifact
val cache = ArchiveCache()
val dir = cache.get(Artifact("https://cosmo.zip/pub/cosmocc/cosmocc-3.9.7.zip"))
  .unsafeRun()(cache.cache.ec).toTry.get
os.list(os.Path(dir))
```

**Comment by alexarchambault** [comment]
For arbitrary post-processing, I wouldn't recommend modifying files or directories that coursier returns, but… some changes should be fine: making a file executable, adding new files in an extracted archive, etc. Once an archive is extracted, coursier doesn't check for added (or removed) files.

**Comment by alexarchambault** [comment]
> I think to make this a reality, a few steps are necessary:
> 
> 1. Write reference documentation for programmatic use of the coursier cache API somewhere. https://get-coursier.io/docs/cache only covers the behavior of the cache but not how to actually use it from code. We could add the API usage docs there, or on mill-build.org, but it needs to live somewhere

I agree more API docs (sometimes, even some API doc) is needed

**Comment by alexarchambault** [comment]
> 2\. Decide on whether to expose Coursier's API as-is or wrap it in a Mill wrapper helper. This depends largely on how well Coursier's Cache API fits the com-lihaoyi style that Mill and it's libraries are written in (simple function calls and objects, lots of default parameters, minimal builder or factory or config objects)

I wouldn't say it fits the com-lihaoyi style, currently. It needs a lot of polishing, at a minimum. Here is how one can enable logging with progress bars, in the archive example above:
```scala
import coursier.cache.{ArchiveCache, FileCache}
import coursier.cache.loggers.RefreshLogger
import coursier.util.Artifact

val cache = FileCache().withLocation((os.pwd / "tmp" / "cache").toIO)
  .withLogger(RefreshLogger.create())
val archiveCache = ArchiveCache()
  .withCache(cache)
  .withLocation((os.pwd / "tmp" / "arc").toIO)
val dir = archiveCache.get(Artifact("https://cosmo.zip/pub/cosmocc/cosmocc-3.9.7.zip"))
  .unsafeRun()(cache.ec).toTry.get
os.list(os.Path(dir))
```

Default values were despised in the first years of coursier (which was actually a good thing in the core of the resolver). Then I started using [data classes](https://github.com/alexarchambault/data-class) a lot, so there's a lot of such data classes around. But there's very little method calls with default values. And coursier has its own IO (that wraps `scala.concurrent.Future`), `coursier.util.Task`, which is useful internally, but less so for end users.

**Comment by lihaoyi** [comment]
zinc rt.jar files is another thing that needs a global cache https://github.com/com-lihaoyi/mill/issues/4164

**Comment by jodersky** [comment]
> Apart from the download (and maybe unpack) kinds of caching, we also need some story for what to do with external tools outside of our control which require cache locations. PIP and NPM are two examples we're facing right now, but I'm sure there'll be others

One compromise I can think of, is to go with a global task type (cf my first comments), but allow some kind of manual "epoch" specification for cache busting, or a way to override directory name (if the task name is changed). It's still not ideal, but would be more forgiving in case of user error. Maybe we could also implement some kind of super clean command, which specifically cleans global caches by name?


e.g. roughly something like this

```scala
def pythonExe = Task.Globa(epoch = 0) { // increase epoch to force clear caches
  os.call("pip", "install", ...)
}
```

or

```scala
def pythonExe = Task.Global(cacheName = os.sub / "foo-1" ) { // cacheName is a relative dir within $CACHE
  os.call("pip", "install", ...)
}
```

`T.dest` would correspond to a directory within a global cache directory (e.g. ~/.cache on linux), in the task. These directories would be kept through `mill clean`, but we could still allow mill to evict them with a separate command, e.g. `mill clean --global` or similar

**Comment by lihaoyi** [comment]
> I wouldn't say it fits the com-lihaoyi style, currently. It needs a lot of polishing, at a minimum. Here is how one can enable logging with progress bars, in the archive example above:

I think we likely will need to have some kind of custom wrapper for Coursier, because Coursier's built in progress logger won't play well with Mill's normal output logger (which only accepts full lines and doesn't allow interactivity), and so we would need to wire up coursier's progress interface into Mill's PromptLogger to show the progress as part of the fixed line-per-task prompt at the bottom of the terminal

**Comment by lihaoyi** [comment]
> def pythonExe = Task.Global(cacheName = os.sub / "foo-1" ) { // cacheName is a relative dir within $CACHE
>   os.call("pip", "install", ...)
> }

I wonder if we should be even stricter and assign cache subfolder directly based on the fully qualified class name of some class representing the cache key? 

* That would mirror what we do for Mill tasks (both in `build.mill`, `package.mill`, and extrernal modules), 
* It would be a bit more verbose than custom `/foo-1/` subfolders, 
* It has the advantage that it's type-safe, e.g. you won't accidentally mis-spell the cache subfolder
* It's less likely to collide: you won't have two independent implementations accidentally using the same cache folder, since they'll probably have different package paths leading up to the fully qualified class name)

**Comment by jodersky** [comment]
> sign cache subfolder directly based on the fully qualified class name of some class representing the cache key

Can you give an example to help me understand what you mean?

I agree that there should be a default. The wording I used in my example wasn't great, what I meant was for a way to *override* a name, so that it's possible to either manually bust previous caches, or keep names for backwards compatibility if needed.

**Comment by lefou** [comment]
So IIUC, we more or less agreed to have some coursier-based URL-keyed cached for immutable content. It should essentially wrap the behavior of `coursier get <URL>` in some Mill API.

Since we have a lot of modules that would profit from such caching already, we should start with a simple task in `CoursierModule`. The parameters should take the URL (as string), an optional known hash (inkl. the hash algorithm) and maybe a TTL). Don't have a good name yet. Maybe `fetchUrl`.

About @sideeffffect 's feature request to be able to feed a list of URLs to external tools. I myself needed this style of downloading a long time ago when I bootstrapped and updated a Gentoo system which was badly connected over a modem to the web. The Gentoo package manager `portage` has some CLI option to only fetch dependencies (`--fetch-only`) and one that only printed the required resources and their locations (I don't remember the correct CLI option). I also see the need for it today, e.g. when setting up Mill in NIX. And also a Gentoo ebuild, that want's to build a Mill-based project from source need such support. 

When we use coursier for most dependency resolution and also for fetching remote resources, an API to get URLs or a list of `cs` invocations is probably not that hard, but would cover a lot of use cases. We could add a new command to `OfflineSupportModule` that returns such a list.

To make remote resources easily detectable, I could imagine a new task type (the name is for debate) `def someResource = Task.RemoteResources(RemoteReosurce(url = ..., hash = ..., ...), ...)` that results in a easy to identify task. It should be easy to collect a list of all such resources from some `prepareOffline` command.

**Comment by alexarchambault** [comment]
I think we could get a nice offline mode the following way: basically we could
- require all downloads to go through the coursier cache
- require the coursier cache instance to come from a specific Mill-defined anon task (like `trait Downloader extends Module { def cache: Task[CoursierCache] = Task.Anon { … } }`)

We could easily spot in the task graph which tasks depend directly on `Downloader#cache`. These are the tasks that download things. Then "preparing for offline" would basically mean evaluate all those tasks (which would ensure everything is in the coursier cache).

With that, no need of ad hoc offline tasks or modules like `OfflineSupportModule`. Offline preparation basically comes for free, as soon as a task depends on `Downloader#cache`.

**Comment by alexarchambault** [comment]
To get a list of URLs before downloading them, I have a slightly less clearer picture of it. We could introduce a `CacheRef` or an `ArtifactRef`, which could be returned by some tasks as a substitute to `PathRef`. Instead of holding a local path, an `ArtifactRef` would hold a URL (and maybe extra attributes, like if the artifact returning 404 should be ignored, or a checksum if users want that to be checked).

These should be returned by tasks downloading things. Mill would be in charge of downloading them if needed, and should offer a way to convert them to `PathRef`s. When a task gets `ArtifactRef`s via its inputs, it could assume the artifact is already in cache, so that the conversion to a `PathRef` is basically instant.

But now, tasks returning `ArtifactRef`s don't always need a coursier cache instance anymore, given they don't need to download things themselves. So to spot tasks downloading things, we'd need to spot beforehand in the task graph which tasks return `ArtifactRef`s in one form or another. I don't know if that's possible, given tasks can return things like `(String, PathRef, ArtifactRef)` or even a case class that has `ArtifactRef`s in some of its fields (or a case class that holds collections of other case classes, that have `ArtifactRef`s in some of their fields, etc.).

Maybe the `ArtifactRef` constructor could be private, and instances of it only created by an `ArtifactRefMaker`, obtained via a dedicated task like the coursier cache above? (And we could spot tasks directly depending on it.)

**Comment by lihaoyi** [comment]
> So, while the whole caching process can be modelled as outlined above, we probably want to make it a bit easier for the user.

> We could introduce a new task type Task.remoteResource(url = ..., hash = ...) or better Task.remoteResource(artifactRef: ArtifactRef*). Additionally to the apply() method, it also needs to have a resourceRefs: Seq[ResourceRef] method to get access to it's parameters.

> It's result could return the downloaded and/or cached resources. But traversing tasks and matching on their type should allow us to get the ResourceRefs without fetching them.

> This means we only need a single task per resource or resources, which can be automatically cached and managed by Mill, including inspecting the URLs without fetching anything.

I think this sounds reasonable. By having it as a task type that makes it possible to collect the callsites by traversing the task graph and module tree, which wouldn't be possible if the callsites were just helper methods.

I expect `remoteResource` or `ArtifactRef` would end up collecting a bunch of flags, since we don't want to allow it to take arbitrary code to allow the hashes to be stable across Mill versions. For example:

- `unpackArchive: Boolean`, for tarballs or zip files which we want to unpack onto disk before caching
- `referencedSubPath: os.SubPath`, for unpacked tarballs or zip files where we don't have to `PathRef` the whole thing which might be expensive if the archive is large
- `downloadHeaders: Iterable[(String, String)]`, for downloads that need special HTTP configuration
- `contentHash: String`, to perform an assert on the integrity of the download and ensure it is byte-for-byte what is expected

I'm not sure what else we may have to put in here, but it seems that the set of things we need to put in here might be relatively finite.

**Comment by lefou** [comment]
I have just another use case, which I'd like to address before we implement this. Sometimes, you need to work with software distributions, e.g. some Eclipse Archives, to extract some jars which you can't find on any Maven repository. Archives, esp. older ones may be distributed to various servers or mirrors which means, their actual download URLs might not be the same, while the actual archive is. In our current design, we use the download URL as cache key, which would mean, we might end up storing the same archive multiple times in the cache if it is downloaded from different sources. 

What are our options:

1. We could provide alternative download urls, but always use the primary url as cache key. We should handle download failure by re-trying with an alternative url as long as we know one.
2. We could specify an explicit cache key to use. This is worse than 1. since we might end up with a cache location, that no longer reveals its origin.
3. If we know the archive hash, we could cache the file under some cache path encoding the hash. Downside to this is, we must known the file hash beforehand and the cache path doesn't reveal its origin.
